src/index/StringSortComparator.h

```diff
@@ -0,0 +1,462 @@
+    setupCollators();
+    setIgnorePunctuationOnFirstLevels(_ignorePunctuationStatus);
+  }
+
+  // Current usage requires default-constructibility. Default to "en_US" and
+  // respecting symbols on all the levels
+  LocaleManager() : LocaleManager("en", "US", false){};
+
+  /**
+   * @param lang The language of the locale, e.g. "en" or "de"
+   * @param country The country of the locale, e.g. "US" or "CA"
+   * @param ignorePunctuationAtFirstLevel If true then spaces/punctuation etc.
+   * will only be considered for comparisons if strings match otherwise Throws
+   * std::runtime_error if the locale cannot be constructed from lang and
+   * country args
+   *
+   * \todo: make the exact punctuation level configurable.
+   */
+  LocaleManager(const std::string& lang, const std::string& country,
+                bool ignorePunctuationAtFirstLevel) {
+    _icuLocale = icu::Locale(lang.c_str(), country.c_str());
+    _ignorePunctuationStatus =
+        ignorePunctuationAtFirstLevel ? UCOL_SHIFTED : UCOL_NON_IGNORABLE;
+
+    if (_icuLocale.isBogus() == TRUE) {
+      throw std::runtime_error("Could not create locale with language " + lang +
+                               " and Country " + country);
+    }
+    setupCollators();
+    setIgnorePunctuationOnFirstLevels(_ignorePunctuationStatus);
+  }
+
+  // ____________________________________________________________________
+  LocaleManager& operator=(const LocaleManager& other) {
+    if (this == &other) return *this;
+    _icuLocale = other._icuLocale;
+    _ignorePunctuationStatus = other._ignorePunctuationStatus;
+    setupCollators();
+    setIgnorePunctuationOnFirstLevels(_ignorePunctuationStatus);
+    return *this;
+  }
+
+  // _____________________________________________________________________
+  [[nodiscard]] int compare(std::string_view a, std::string_view b,
+                            const Level level) const {
+    UErrorCode err = U_ZERO_ERROR;
+    auto idx = static_cast<uint8_t>(level);
+    auto res = compToInd(
+        _collator[idx]->compareUTF8(toStringPiece(a), toStringPiece(b), err));
+    raise(err);
+    return res;
+  }
+
+  // _____________________________________________________________________
+  static int compare(WeightString a, WeightString b,
+                     [[maybe_unused]] const Level) {
+    return std::strcmp(a.get().c_str(), b.get().c_str());
+  }
+
+  // transform a UTF-8 string into a weight string that can be compared using
+  // std::strcmp. needed because ICU only works on utf16 and does not create a
+  // std::string.
+  [[nodiscard]] WeightString getSortKey(std::string_view s,
+                                        const Level level) const {
+    auto utf16 = icu::UnicodeString::fromUTF8(toStringPiece(s));
+    auto& col = *_collator[static_cast<uint8_t>(level)];
+    auto sz = col.getSortKey(utf16, nullptr, 0);
+    WeightString finalRes;
+    std::string& res = finalRes.get();
+    res.resize(sz);
```

Comment:
Are we absolutely sure we have to convert to UTF-16 and ICU really is that shitty?
http://userguide.icu-project.org/strings/utf-8 sounds like they have at least some functionality to directly work with UTF-8 and in particular the [`Collator::compareUTF8()`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1Collator.html#a58f0ddc2202e376cc912015fa56fd42f) function sounds like it *should* be enough!?

Reply:
For simple comparisons between strings, the `compareUTF8()` method is fine and we use it (e.g. when initially sorting the vocabulary). However we also need to properly answer the query "is `a` a prefix of `b`" which is non-trivial to do correctly as I have found:
Approach A: Get a prefix of the longer string of appropriate length and compare for equality. Problem: Since we might ignore certain characters on certain levels (e.g. punctuations or combined diacritics) it is not really possible to estimate the correct prefix length without inspecting each codepoint with manually specified rules which again defeats the purpose of ICU.
Approach B: Implement a `comparePrefix` method in ICU. This is currently clearly out of scope for me especially since the legacy ICU codebase does not seem to use/allow templates. I am think about starting a discussion on their github, but this is currently beyond our scope.
Approach C: Hacking the SortKeys. This is basically the generalized version of "every string that is >= "abc" and <"abd" has the prefix "abc". However this currently only works on the primary level, that's why I hardcoded it by basically increasing the last byte of the sortKey by one. Since I now know how the ICU sortKeys are created (A special `\1` byte is used as a separator between the levels). I will extend this to arbitrary levels.

Unfortunately the `getSortKey`-API of ICU currently only supports UTF-16. I also don't really like this approach, 
but since we are only doing this in binary searches, we should be fine (and since our vocabulary is static we could precalculate the sort keys of the elements that are constantly hit by the binary search if we find the performance to be too poor [i doubt this, but i still have to measure])

Reply:
Thank you for the great explanation! I think opening a discussion with the ICU community would make a lot of sense. I wouldn't put too much priority on extending this to other levels. I suspect that for our main usage of auto completion `PRIMARY` may be exactly what we want. In fact I'm a bit on the fence about allowing the user to configure the sort locale at all. We can't really test with a lot of locales and it seems this is a [mess](https://stackoverflow.com/questions/38961492/how-do-you-set-the-collation-for-a-sparql-query?rq=1) with other SPARQL engines too. So I'm wondering if a sane default that is consistent may be preferable to a lot of new variables.

Reply:
Well as I see it
1.  I agree with you on `PRIMARY` being a useful default for Autocompletion.
2. Concerning configurable LOCALES I am not sure. Maybe we could emit a warning when something else than "en-US" was selected, or require some `--expert-mode` or `--unsafe` flag.
It is somewhat the charm of `ICU` to have every aspect of the locales configurable.

Reply:
I like the idea of a warning, doesn't have to be stern just a reminder that any other locale isn't well tested and needs to be mentioned when reporting an issue. Also it should mention that the locale is fixed for the lifetime of the index and can only be changed by regenerating.